runtimes/runtimes/lsp/router/lspRouter.ts

```diff
@@ -65,4 +81,31 @@ export class LspRouter {
             }
         }
     }
+
+    getUserAgent = (params?: AWSClientInitializationOptions): string => {
+        const items: String[] = []
+
+        // Standard prefix for all Language Server Runtimes artifacts
+        items.push(USER_AGENT_PREFIX)
+
+        // Fields, specific to runtime artifact
+        items.push(`${this.name}/${this.version}`)
+
+        // Compute client-specific suffix
+        const { product, platform, clientId } = params || {}
+
+        if (product?.name && product?.version) {
+            items.push(`${product.name}/${product.version}`)
+        }
+
+        if (platform?.name && platform?.version) {
+            items.push(`${platform.name}/${platform.version}`)
+        }
```

Comment:
If name or version is missing but the other is present, you should still emit the data you do have. I suggest replacing the missing value with some token like "UNKNOWN" or "MISSING"

Reply:
On second thought, you may want to check all fields for null and swap them, so you don't emit something like "USER_AGENT_PREFIX /1.2.3" (in this example this.version is present but this.name is absent). I assume this would trip up any parsers or queries consuming the agent.

Reply:
should we just let the client build this and pass it as-is?

Reply:
> should we just let the client build this and pass it as-is?

My idea with building it in runtime is to have a single source of truth for the format based on expected input, so it is predictable.

I considered making clients responsible for passing the suffix, which will make this implementation much simpler.

Reply:
I revised proposed protocol, and simplified it to pass only userAgentSuffix from the client. Some data will still be attached in Runtime, but logic is simpler

Reply:
> should we just let the client build this and pass it as-is?

I'm concerned if user agent strings need to be in a specific format. This wont scale well, because each client may/will diverge, and then there will be change requests to fix this in both clients and the back-end as mistakes occur in what is passed along as a pre-built string.